pkg/controllers/cliconfig/cli_config_controller.go

```diff
@@ -0,0 +1,472 @@
+	routev1lister "github.com/openshift/client-go/route/listers/route/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/configobservation"
+)
+
+var (
+	scheme  = runtime.NewScheme()
+	codecs  = serializer.NewCodecFactory(scheme)
+	encoder = codecs.LegacyCodec(osinv1.GroupVersion) // TODO I think there is a better way to do this
+)
+
+func init() {
+	utilruntime.Must(osinv1.Install(scheme))
+}
+
+// knownConditionNames lists all condition types used by this controller.
+// These conditions are operated and defaulted by this controller.
+// Any new condition used by this controller sync() loop should be listed here.
+var knownConditionNames = sets.NewString(
+	"OAuthConfigDegraded",
+	"OAuthSessionSecretDegraded",
+	"OAuthServerIngressDegraded",
+	"OAuthServerServiceDegraded",
+	"AuthCLIConfigProgressing",
+)
+
+type cliConfigController struct {
+	services corev1lister.ServiceLister
+	route    routev1lister.RouteLister
+
+	auth           operatorv1client.AuthenticationsGetter
+	configMaps     corev1client.ConfigMapsGetter
+	secrets        corev1client.SecretsGetter
+	operatorClient v1helpers.OperatorClient
+}
+
+func NewCLIConfigController(kubeInformersForTargetNamespace informers.SharedInformerFactory, secrets corev1client.SecretsGetter, configMaps corev1client.ConfigMapsGetter,
+	operatorClient v1helpers.OperatorClient, authentication operatorv1client.AuthenticationsGetter, routeInformer routeinformer.RouteInformer, recorder events.Recorder) factory.Controller {
+	c := &cliConfigController{
+		services:       kubeInformersForTargetNamespace.Core().V1().Services().Lister(),
+		route:          routeInformer.Lister(),
+		secrets:        secrets,
+		configMaps:     configMaps,
+		operatorClient: operatorClient,
+		auth:           authentication,
+	}
+	return factory.New().WithInformers(
+		kubeInformersForTargetNamespace.Core().V1().Secrets().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().Services().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().ConfigMaps().Informer(),
+		routeInformer.Informer(),
+		operatorClient.Informer(),
+	).ResyncEvery(30*time.Second).WithSync(c.sync).ToController("CLIConfig", recorder.WithComponentSuffix("cli-config-controller"))
+}
+
+func (c *cliConfigController) getRoute() (*routev1.Route, []operatorv1.OperatorCondition) {
+	// route is a pre-requirement for this sync
+	// it is created by metadata controller
+	// if route does not exists, do nothing and wait
+	route, err := c.route.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "OAuthServerIngressDegraded",
```

Comment:
Will every controller have its own "couldn't get route" condition now? Should it be something more generic, like `type: "CLIConfigControllerSyncDegraded"` which would cover most of these cases?

Reply:
if just one controller get stuck, then I would rather prefer each controller having its own condition.